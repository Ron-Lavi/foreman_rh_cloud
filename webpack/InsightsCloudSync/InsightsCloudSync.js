import React, { useEffect } from 'react';
import { Button } from 'patternfly-react';
import PropTypes from 'prop-types';
import { translate as __ } from 'foremanReact/common/I18n';
import PageLayout from 'foremanReact/routes/common/PageLayout/PageLayout';
import { useForemanSettings } from 'foremanReact/Root/Context/ForemanContext';
import InsightsHeader from './Components/InsightsHeader';
import {
  INSIGHTS_SYNC_PAGE_TITLE,
  INSIGHTS_SEARCH_PROPS,
} from './InsightsCloudSyncConstants';

const InsightsCloudSync = ({
  syncInsights,
  updateUrlQuery,
  query,
  fetchInsights,
  page,
  perPage: urlPerPage,
}) => {
  const { perPage: appPerPage } = useForemanSettings();
  const perPage = urlPerPage || appPerPage;
  useEffect(() => {
    fetchInsights({ page, perPage, query });
  }, [fetchInsights, page, perPage, query]);
  return (
    <PageLayout
      searchable
      searchProps={INSIGHTS_SEARCH_PROPS}
      onSearch={searchQuery => updateUrlQuery({ searchQuery, page: 1 })}
      header={INSIGHTS_SYNC_PAGE_TITLE}
      toolbarButtons={
        <Button bsStyle="primary" onClick={syncInsights}>
          {__('Sync now')}
        </Button>
      }
      searchQuery={query}
      beforeToolbarComponent={<InsightsHeader />}
    >
      <React.Fragment>
        <p>Insights Table will be here</p>
      </React.Fragment>
    </PageLayout>
  );
};

InsightsCloudSync.propTypes = {
  syncInsights: PropTypes.func.isRequired,
  updateUrlQuery: PropTypes.func.isRequired,
  fetchInsights: PropTypes.func.isRequired,
  page: PropTypes.number,
  perPage: PropTypes.number,
  query: PropTypes.string,
};

InsightsCloudSync.defaultProps = {
  page: 1,
  perPage: null,
  query: '',
};

export default InsightsCloudSync;
