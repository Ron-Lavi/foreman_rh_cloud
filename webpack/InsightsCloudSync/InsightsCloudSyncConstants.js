import { translate as __ } from 'foremanReact/common/I18n';
import { getControllerSearchProps } from 'foremanReact/constants';
import { foremanUrl } from '../ForemanRhCloudHelpers';

export const INSIGHTS_CLOUD_SYNC_SUCCESS = 'INSIGHTS_CLOUD_SYNC_SUCCESS';
export const INSIGHTS_SYNC_PAGE_TITLE = __('Red Hat Insights');
export const INSIGHTS_PATH = foremanUrl('/foreman_rh_cloud/insights_cloud');
export const INSIGHTS_HITS_PATH = '/insights_cloud/hits';
export const INSIGHTS_HITS_API_KEY = 'INSIGHTS_HITS';
export const INSIGHTS_SEARCH_PROPS = {
  ...getControllerSearchProps('/insights_cloud/hits'),
  controller: INSIGHTS_PATH,
};
