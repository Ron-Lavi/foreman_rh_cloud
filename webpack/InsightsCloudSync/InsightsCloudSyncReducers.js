import { combineReducers } from 'redux';
import Immutable from 'seamless-immutable';
import { reducers as settingsReducers } from './Components/InsightsSettings';

const initialState = Immutable({ hits: [], itemCount: 0, selectedIds: {} });

export const tableReducer = (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case 'SET_SELECTED_IDS':
      return state.merge(payload);
    default:
      return state;
  }
};

export default {
  InsightsCloudSync: combineReducers({
    ...settingsReducers,
    table: tableReducer,
  }),
};
