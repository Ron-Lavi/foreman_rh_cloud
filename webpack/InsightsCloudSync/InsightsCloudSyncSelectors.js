import URI from 'urijs';
import { selectRouterLocation } from 'foremanReact/routes/RouterSelector';

export const selectQuery = state => selectRouterLocation(state).query;

export const selectSearch = state => {
  const { search = '' } = selectQuery(state);
  return URI.decodeQuery(search);
};

export const selectPage = state => {
  const { page = '1' } = selectQuery(state);
  return Number(page);
};

export const selectPerPage = state => {
  const perPage = selectQuery(state).per_page;
  return perPage && Number(perPage);
};

export const selectQueryParams = state => ({
  page: selectPage(state),
  perPage: selectPerPage(state),
  searchQuery: selectSearch(state),
});
