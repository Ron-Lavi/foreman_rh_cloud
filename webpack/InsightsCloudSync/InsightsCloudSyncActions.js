import { push } from 'connected-react-router';
import API from 'foremanReact/API';
import { get } from 'foremanReact/redux/API';
import { addToast } from 'foremanReact/redux/actions/toasts';
import { stringifyParams } from 'foremanReact/common/urlHelpers';
import { insightsCloudUrl } from './InsightsCloudSyncHelpers';
import { selectQueryParams } from './InsightsCloudSyncSelectors';
import {
  INSIGHTS_CLOUD_SYNC_SUCCESS,
  INSIGHTS_HITS_PATH,
  INSIGHTS_HITS_API_KEY,
  INSIGHTS_PATH,
} from './InsightsCloudSyncConstants';

export const syncInsights = () => async dispatch => {
  try {
    await API.post(insightsCloudUrl('tasks'));
    dispatch({
      type: INSIGHTS_CLOUD_SYNC_SUCCESS,
      payload: {},
    });
  } catch ({ message }) {
    dispatch(
      addToast({
        sticky: true,
        type: 'error',
        message,
      })
    );
  }
};

export const updateUrlQuery = (queryParams = {}) => (dispatch, getState) => {
  dispatch(
    push({
      pathname: INSIGHTS_PATH,
      search: stringifyParams({
        ...selectQueryParams(getState()),
        ...queryParams,
      }),
    })
  );
};

export const fetchInsights = ({ page, perPage, query, sort } = {}) => {
  const order =
    sort && Object.keys(sort).length > 0 ? `${sort.by} ${sort.order}` : '';
  return get({
    key: INSIGHTS_HITS_API_KEY,
    url: INSIGHTS_HITS_PATH,
    params: {
      page,
      per_page: perPage,
      search: query,
      order,
    },
  });
};
