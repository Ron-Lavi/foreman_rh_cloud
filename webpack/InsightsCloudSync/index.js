import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './InsightsCloudSyncActions';
import InsightsCloudSync from './InsightsCloudSync';
import insightsCloudSyncReducers from './InsightsCloudSyncReducers';
import {
  selectPage,
  selectPerPage,
  selectSearch,
} from './InsightsCloudSyncSelectors';

// map state to props
const mapStateToProps = state => ({
  query: selectSearch(state),
  page: selectPage(state),
  perPage: selectPerPage(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = insightsCloudSyncReducers;

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(InsightsCloudSync);
