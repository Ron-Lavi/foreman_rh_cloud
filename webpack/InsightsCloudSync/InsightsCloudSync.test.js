import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import { noop } from 'patternfly-react';

import InsightsCloudSync from './InsightsCloudSync';

const fixtures = {
  render: {
    syncInsights: noop,
    updateUrlQuery: noop,
    query: '',
    fetchInsights: noop,
  },
};

describe('InsightsCloudSync', () =>
  testComponentSnapshotsWithFixtures(InsightsCloudSync, fixtures));
