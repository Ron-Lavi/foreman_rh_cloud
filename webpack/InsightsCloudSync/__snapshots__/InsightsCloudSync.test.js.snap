// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsightsCloudSync render 1`] = `
<PageLayout
  beforeToolbarComponent={<InsightsHeader />}
  header="Red Hat Insights"
  onSearch={[Function]}
  searchProps={
    Object {
      "autocomplete": Object {
        "id": "searchBar",
        "searchQuery": "",
        "url": "/insights_cloud/hits/auto_complete_search",
        "useKeyShortcuts": true,
      },
      "bookmarks": Object {
        "canCreateBookmarks": true,
        "documentationUrl": "4.1.5Searching",
        "url": "/api/bookmarks",
      },
      "controller": "/foreman_rh_cloud/insights_cloud",
    }
  }
  searchQuery=""
  searchable={true}
  toolbarButtons={
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Sync now
    </Button>
  }
>
  <p>
    Insights Table will be here
  </p>
</PageLayout>
`;
